cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaReconciler.java

```diff
@@ -1040,6 +1057,40 @@ protected Future<Void> updateKafkaVersion(KafkaStatus kafkaStatus) {
         return Future.succeededFuture();
     }
 
+    protected Future<Void> updateKafkaMetadataState(KafkaStatus kafkaStatus) {
+        kafkaStatus.setKafkaMetadataState(kafkaMetadataStateManager.computeNextMetadataState().name());
+        return Future.succeededFuture();
+    }
+
+    protected Future<Void> maybeKafkaMetadataMigrationInProgress() {
+        KafkaMetadataStateManager.KafkaMetadataConfigurationState kafkaMetadataConfigState = this.kafkaMetadataStateManager.computeMetadataConfigurationState();
+        // on each reconcile, would be useless to check migration status if it's not going on
+        if (kafkaMetadataConfigState.equals(KafkaMetadataStateManager.KafkaMetadataConfigurationState.MIGRATION)) {
+            LOGGER.infoCr(reconciliation, "Checking ZooKeeper migration state");
+            return ReconcilerUtils.clientSecrets(reconciliation, secretOperator)
+                    .compose(compositeFuture -> {
+                        // we should get the quorum controller leader using the Admin Client API describeMetadataQuorum() but
+                        // it's not supported by brokers during migration because they are still connected to ZooKeeper so ...
+                        // picking a random controller from the set, because all expose the needed metrics
+                        NodeRef[] controllers = kafka.controllerNodes().toArray(new NodeRef[kafka.controllerNodes().size()]);
+                        int controllerId = controllers[RANDOM.nextInt(kafka.controllerNodes().size())].nodeId();
```

Comment:
It was a way to avoid picking the same controller consistently if there is any kind of problem with it rolling, restarting or any other problem and not able to provide the metrics we need. Getting the leader would be the best solution but as per comment above it's not possible. Why do you think it doesn't make any sense?

Reply:
Well, it is not consistent when it is random. It is the opposite, irreproducible. If the first node has a problem, I guess I would also wonder what it really means for the cluster. But also, the likelihood of the second or third node having problems is the same as the first one. 

I think for example having a `while` loop over all nodes until you find the answer would be a better approach than the random one and it would also have better guarantees of getting the information from at least some of the nodes. And it would save us keeping the Randon object in the class as well.

Reply:
I have refactored using a loop, let me know what you think please.